processor/src/main/java/org/mapstruct/ap/internal/model/AdditionalAnnotationsBuilder.java

```diff
@@ -90,6 +92,44 @@ protected void addInstances(AnnotateWithsGem gem, Element source, Set<Annotation
         }
     }
 
+    @Override
+    public Set<Annotation> getProcessedAnnotations(Element source) {
+        Set<Annotation> processedAnnotations = super.getProcessedAnnotations( source );
+        return addDeprecatedAnnotation( source, processedAnnotations );
+    }
+
+    private Set<Annotation> addDeprecatedAnnotation(Element source, Set<Annotation> annotations) {
+        DeprecatedGem deprecatedGem = DeprecatedGem.instanceOn( source );
+        if ( deprecatedGem == null ) {
+            return annotations;
+        }
+        Type deprecatedType = typeFactory.getType( Deprecated.class );
+        if ( annotations.stream().anyMatch( annotation -> annotation.getType().equals( deprecatedType ) ) ) {
+            messager.printMessage(
+                    source,
+                    deprecatedGem.mirror(),
+                    Message.ANNOTATE_WITH_DUPLICATE,
+                    deprecatedType.describe() );
+            return annotations;
+        }
+        List<AnnotationElement> annotationElements = new ArrayList<>();
+        for ( Map.Entry<? extends ExecutableElement, ? extends AnnotationValue> entry : deprecatedGem.mirror()
```

Comment:
We don't really need to use the mirror and strings for this. The `DeprecatedGem` should have methods that allow to have access to the values. You can rewrite the loop to:

```java
if ( deprecatedGem.since().hasValue() ) {
    annotationElements.add( new AnnotationElement(
        AnnotationElementType.STRING,
        "since",
        Collections.singletonList( deprecatedGem.since().getValue() )
    ) );
}
        
if ( deprecatedGem.forRemoval().hasValue() ) {
    annotationElements.add( new AnnotationElement(
        AnnotationElementType.BOOLEAN,
        "forRemoval",
        Collections.singletonList( deprecatedGem.forRemoval().getValue() )
    ) );
}
```